lib/pet_store/accounts/user.ex

```diff
@@ -60,6 +62,15 @@ defmodule PetStore.Accounts.User do
     |> maybe_hash_password(opts)
   end
 
+  defp validate_admin_level(changeset, opts) do
+    changeset
+    |> validate_inclusion(:admin_level, 0..5, message: "invalid range")
+    |> validate_number(:admin_level,
+      less_than_or_equal_to: Keyword.get(opts, :max_admin_level, 5),
+      message: "insufficient permissions"
+    )
+  end
```

Comment:
My 2 cents: you're mixing validation (i.e. "Is the `admin_level` in a valid range) with business logic rule enforcement ("Is the creator allowed to do this"). The former belongs to changesets, the latter does not.
This is also the reason why in the controller you then have to pattern match on a specific error in the changeset to transform it in `{:error, :forbidden}`. If you perform the the authz at the context level, you can just return `{:error, :forbidden}` from there.
Remember that you can always do `Ecto.Changeset.apply_changes` in the context to validate business logic rules before hitting the database

Reply:
Requested changes applied in https://github.com/noaccOS/Pet-Store/pull/4/commits/8d32459b40b133891c80259c1d89e1a553c8423c

Reply:
> Remember that you can always do `Ecto.Changeset.apply_changes` in the context to validate business logic rules before hitting the database

I think you mean `apply_action/2`?

https://hexdocs.pm/ecto/Ecto.Changeset.html#apply_changes/1
https://hexdocs.pm/ecto/Ecto.Changeset.html#apply_action/2